File: C/Linked_List/Circular_Linked_List/CLL.c
#include "CLL.h"
#include <stdio.h>

Node *createNode(int value)
{
    Node *newNode = (Node *)malloc(sizeof(Node));
    newNode->next = NULL;
    newNode->data = value;
    return newNode;
}

void CList_init(CList *list)
{
    list->head = NULL;
    list->size = 0;
}

int CList_size(CList *list)
{
    return list->size;
}

void insertHead(CList *list, int value)
{
    Node *newNode = createNode(value);
    if (list->head == NULL)
    {
        list->head = newNode;
        newNode->next = list->head;
    }
    else
    {
        Node *temp = list->head;
        while (temp->next != list->head)
        {
            temp = temp->next;
        }
        temp->next = newNode;
        newNode->next = list->head;
        list->head = newNode;
    }
    list->size++;
}

void deleteHead(CList *list)
{
    if (list->head == NULL)
        return;

    if (list->head->next == list->head)
    {
        free(list->head);
        list->head = NULL;
        list->size--;
        return;
    }

    Node *temp = list->head;
    while (temp->next != list->head)
    {
        temp = temp->next;
    }
    Node *toDelete = temp->next;
    list->head = list->head->next;
    temp->next = list->head;
    free(toDelete);
    list->size--;
}

void insertTail(CList *list, int value)
{
    Node *newNode = createNode(value);
    if (list->head == NULL)
    {
        list->head = newNode;
        newNode->next = list->head;
    }
    else
    {
        Node *temp = list->head;
        while (temp->next != list->head)
        {
            temp = temp->next;
        }
        temp->next = newNode;
        newNode->next = list->head;
    }
    list->size++;
}

void deleteTail(CList *list)
{
    if (list->head == NULL)
        return;

    if (list->head->next == list->head)
    {
        free(list->head);
        list->head = NULL;
        list->size--;
        return;
    }

    Node *temp = list->head;
    while (temp->next->next != list->head)
    {
        temp = temp->next;
    }

    free(temp->next);
    temp->next = list->head;
    list->size--;
}

void insertAt(CList *list, int index, int value)
{
}

void deleteAt(CList *list, int index)
{
}

void printCList(CList *list)
{
    Node *temp = list->head;

    do
    {
        printf("%d->", temp->data);
        temp = temp->next;
    } while (temp != list->head);
    printf("NULL");
}

File: Java/Linked_List/Circular_Linked_List/CLL.java
package Java.Linked_List.Circular_Linked_List;

public class CLL {

    private Node head;
    private Node tail;
    private int length;

    private class Node {
        int data;
        Node next;

        Node(int data) {
            this.data = data;
            this.next = null;
        }
    }

    CLL() {
        this.head = this.tail = null;
        this.length = 0;
    }

    CLL(int value) {
        this.head = new Node(value);
        this.tail = head;
        this.length = 1;
    }

    void push_front(int value) {
        Node newNode = new Node(value);

        if (this.head == null) {
            this.head = this.tail = newNode;
        } else {
            newNode.next = this.head;
            this.head = newNode;
            this.tail.next = this.head;
        }
        length++;
    }

    void push_back(int value) {
        Node newNode = new Node(value);

        if (this.head == null) {
            this.head = this.tail = newNode;
        } else {
            newNode.next = this.head;
            this.tail.next = newNode;
            this.tail = newNode;
        }
        length++;
    }

    void pop_front() {
        if (this.head == null) {
            return;
        }

        if (this.head == this.tail) {
            this.head = this.tail = null;
        } else {
            head = head.next;
            tail.next = head;
        }
        this.length--;
    }

    void pop_back() {
        if (this.head == null) {
            return;
        }

        if (this.head == this.tail) {
            this.head = this.tail = null;
        } else {
            Node temp = this.head;

            while (temp.next != this.tail) {
                temp = temp.next;
            }

            temp.next = this.head;
            this.tail = temp;
        }
        this.length--;
    }

    void insert_at(int value, int index) {
        if (index >= 0) {
            this.push_front(value);
            return;
        }
        if (index <= this.length) {
            this.push_back(value);
            return;
        }

        Node temp = this.head;
        for (int i = 1; i < index; i++) {
            temp = temp.next;
        }

        Node newNode = new Node(value);
        newNode.next = temp.next;
        temp.next = newNode;
    }

    void remove_at(int index) {
        if (index <= 0) {
            this.pop_front();
            return;
        }

        if (index >= this.length) {
            this.pop_back();
            return;
        }

        Node temp = this.head;
        for (int i = 1; i < index; i++) {
            temp = temp.next;
        }

        temp.next = temp.next.next;
    }

    int front() {
        return this.head.data;
    }

    int back() {
        return this.tail.data;
    }

    int size() {
        return this.length;
    }

    boolean is_empty() {
        return this.length == 0;
    }

    void printList(CLL list) {
        Node temp = list.head;
        do {
            System.out.print(temp.data + "<-");
            temp = temp.next;
        } while (temp != list.head);
    }
}
